mm2src/coins/nft/storage/sql_storage.rs

```diff
@@ -231,16 +233,16 @@ fn insert_nft_in_list_sql(chain: &Chain) -> MmResult<String, SqlError> {
     Ok(sql)
 }
 
-fn insert_tx_in_history_sql(chain: &Chain) -> MmResult<String, SqlError> {
-    let table_name = nft_tx_history_table_name(chain);
+fn insert_transfer_in_history_sql(chain: &Chain) -> MmResult<String, SqlError> {
+    let table_name = nft_transfer_history_table_name(chain);
     validate_table_name(&table_name)?;
 
     let sql = format!(
```

Comment:
This is a very dangerous way of building an sql statement and likely subject to sql injection. This goes for all the similarly built sql statements within this file.

Reply:
Oh my mistake, I misread this. Realizing now this is only inserting the `table_name`, not all the parameters.

Reply:
This function is just to build sql command string, which `rusqlite` actually  asks to provide for [`execute()`](https://github.com/KomodoPlatform/komodo-defi-framework/blob/dev/mm2src/coins/nft/storage/sql_storage.rs#L782C55-L782C55)  function for example. 
It requires params `sql: &str, params: P` where in rusqlite, `P: Params` is a trait bound that indicates that the params argument must implement the Params trait. The Params trait is implemented for types that can be used as parameters in a prepared statement. In other words, P: Params means that the params variable must be of a type that can be used as parameters in the execute function. The Params trait is implemented for various types, including tuples, arrays, and slices of types that implement the ToSql trait. 
`execute()` prepares sql statement, binds the parameters to the statement and execute it if you take a look at the source code. You can do it easily if you open project in Clion for example.
![image](https://github.com/KomodoPlatform/komodo-defi-framework/assets/52405288/af983081-530b-4357-84a2-ae80da516f80)


For params we create an array from `Nft` and `NftTransferHistory` fields

Reply:
as for table name, we validate it with `validate_table_name` function which does provide protection against sql injection attacks by only allowing alphanumeric characters and underscores in table names.

Reply:
I'm curious to know if it is justified to store both the individual fields and the entire JSON string containing those fields separately in the table. I'm not sure if there's a better way to do it, but is this approach architecturally justified?

Reply:
> I'm curious to know if it is justified to store both the individual fields and the entire JSON string containing those fields separately in the table. I'm not sure if there's a better way to do it, but is this approach architecturally justified?

Yeah, fair remark. this was originally done because there weren't many filters to get the data, so it was enough to get one field. But now the number of fields for which we receive data or which require changes has grown, so it's worth trying to change the form of the table.

main changes require such functions as `.._from_row` functions like [this](https://github.com/KomodoPlatform/komodo-defi-framework/blob/dev/mm2src/coins/nft/storage/sql_storage.rs#L201) and `finalize_..` like [this](https://github.com/KomodoPlatform/komodo-defi-framework/blob/dev/mm2src/coins/nft/storage/sql_storage.rs#L201).

I noticed that I need to create additional action like `get nft by smth` to update specific field and save correct `details_json`.

Reply:
upd: But also want to make a note, that if we decide to keep fields separately it will require to insert huge amount of params.
There are Nft fields for example
```rust
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct NftCommon {
    pub(crate) token_address: Address,
    pub(crate) token_id: BigDecimal,
    pub(crate) amount: BigDecimal,
    pub(crate) owner_of: Address,
    pub(crate) token_hash: Option<String>,
    #[serde(rename = "name")]
    pub(crate) collection_name: Option<String>,
    pub(crate) symbol: Option<String>,
    pub(crate) token_uri: Option<String>,
    pub(crate) metadata: Option<String>,
    pub(crate) last_token_uri_sync: Option<String>,
    pub(crate) last_metadata_sync: Option<String>,
    pub(crate) minter_address: Option<String>,
    #[serde(default)]
    pub(crate) possible_spam: bool,
    #[serde(default)]
    pub(crate) possible_phishing: bool,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Nft {
    #[serde(flatten)]
    pub(crate) common: NftCommon,
    pub(crate) chain: Chain,
    pub(crate) block_number_minted: Option<u64>,
    pub(crate) block_number: u64,
    pub(crate) contract_type: ContractType,
    pub(crate) uri_meta: UriMeta,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub(crate) struct UriMeta {
    #[serde(rename = "image")]
    pub(crate) raw_image_url: Option<String>,
    pub(crate) image_url: Option<String>,
    #[serde(rename = "name")]
    pub(crate) token_name: Option<String>,
    pub(crate) description: Option<String>,
    pub(crate) attributes: Option<Json>,
    pub(crate) animation_url: Option<String>,
    pub(crate) external_url: Option<String>,
    pub(crate) image_details: Option<Json>,
}
```

Reply:
By the way, I am not very knowledgeable in SQL. However, I have noticed that we are currently using the COUNT(*) function to obtain the total number of entries in an SQLite table. We then use OFFSET and LIMIT to retrieve the desired page in the `finalize_%something%_sql_builder` function. This involves two separate queries. However, some tables have a `log_index` field which could potentially be used as a cursor. In other words, we can retrieve the desired page using the following query:

```
SELECT * 
FROM table_name 
WHERE log_index > %index_of_last_page_end%
ORDER BY log_index 
LIMIT %page_size%;
```

This would only require one query. It is possible that this approach may not be suitable in this case, or the number of entries in such tables may be small enough that there would be no significant difference between the COUNT(*) + OFFSET/LIMIT approach and the approach mentioned above. However, for large datasets, the approach described above would likely be more efficient. 

Just thinking out loud, nothing more.

Reply:
@DeckerSU to be more clear.
Unique `transaction_hash` can have only one transaction. A single transaction can trigger multiple events, and each event generates a log entry. For example, if a single transaction triggers three events, there will be three log entries associated with that transaction, each with a different log_index: 0, 1, and 2.
`log_index` as a primary key was added in this [PR](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1926). If user sends a [batch transfer (see safeBatchTransferFrom)](https://eips.ethereum.org/EIPS/eip-1155#specification) with [multiple NFTs](https://github.com/KomodoPlatform/komodo-wallet/issues/1239#issuecomment-1658670188) you will see in response from moralis transactions with the same `transaction_hash` but different `log_index` for each NFT which was sent in this transaction.

Reply:
@laruh Thank you for the detailed explanation. I now understand that I misunderstood the purpose/function of the `log_index` field. I mistakenly thought it was used for indexing/ordering in the table, but now it is clear that it serves a different purpose. I apologize for taking up your time.

Reply:
> @laruh Thank you for the detailed explanation. I now understand that I misunderstood the purpose/function of the `log_index` field. I mistakenly thought it was used for indexing/ordering in the table, but now it is clear that it serves a different purpose. I apologize for taking up your time.

No no, it alright, we are all here to understand and evaluate each other's code. Please feel free to ask anything! 
Moreover, the question about `details_json` was important, I made a decision to try a bit another version of db tables, where we keep all unchangeable and unnecessary for `select` fields in `details_json`, while other fields will be kept in separate db fields (actually the same tables view as we have now, but another logic for `details_json`). This way should make the `update some specific field` process easier, bcz we dont need anymore to get rust structures from db first, update them and update db tables next. 
Although it can affect get process bcz we have to fetch all fields from db to get rust structure, but code maintenance should get easier.